File: src/core/fiscalyear.ts
import type { UnixTimeRange } from './datetime.ts';

export class FiscalYearStart {
    public static readonly JanuaryFirstDay = new FiscalYearStart(1, 1);
    public static readonly Default = FiscalYearStart.JanuaryFirstDay;

    private static readonly MONTH_MAX_DAYS: number[] = [
        31, // January
        28, // February (Disallow fiscal year start on leap day)
        31, // March
        30, // April
        31, // May
        30, // June
        31, // July
        31, // August
        30, // September
        31, // October
        30, // November
        31 // December
    ];

    public readonly month: number; // 1-based (1 = January, 12 = December)
    public readonly day: number;
    public readonly value: number;

    private constructor(month: number, day: number) {
        this.month = month;
        this.day = day;
        this.value = (month << 8) | day;
    }

    public static of(month: number, day: number): FiscalYearStart | undefined {
        if (!FiscalYearStart.isValidFiscalYearMonthDay(month, day)) {
            return undefined;
        }

        return new FiscalYearStart(month, day);
    }

    /**
     * Create a FiscalYearStart from a uint16 value (two bytes - month high, day low)
     * @param value uint16 value (month in high byte, day in low byte)
     * @returns FiscalYearStart instance or undefined if the value is out of range
     */
    public static valueOf(value: number): FiscalYearStart | undefined {
        if (value < 0x0101 || value > 0x0C1F) {
            return undefined;
        }

        const month = (value >> 8) & 0xFF;  // high byte
        const day = value & 0xFF;           // low byte

        return FiscalYearStart.of(month, day);
    }

    /**
     * Create a FiscalYearStart from a month/day string
     * @param monthDay MM-dd string (e.g. "04-01" = 1 April)
     * @returns FiscalYearStart instance or undefined if the monthDay is invalid
     */
    public static parse(monthDay: string): FiscalYearStart | undefined {
        if (!monthDay || !monthDay.includes('-')) {
            return undefined;
        }

        const parts = monthDay.split('-');

        if (parts.length !== 2) {
            return undefined;
        }

        const month = parseInt(parts[0], 10);
        const day = parseInt(parts[1], 10);

        return FiscalYearStart.of(month, day);
    }

    public toMonthDashDayString(): string {
        return `${this.month.toString().padStart(2, '0')}-${this.day.toString().padStart(2, '0')}`;
    }

    private static isValidFiscalYearMonthDay(month: number, day: number): boolean {
        return 1 <= month && month <= 12 && 1 <= day && day <= FiscalYearStart.MONTH_MAX_DAYS[month - 1];
    }
}

export class FiscalYearUnixTime implements UnixTimeRange {
    public readonly year: number;
    public readonly minUnixTime: number;
    public readonly maxUnixTime: number;

    private constructor(fiscalYear: number, minUnixTime: number, maxUnixTime: number) {
        this.year = fiscalYear;
        this.minUnixTime = minUnixTime;
        this.maxUnixTime = maxUnixTime;
    }

    public static of(fiscalYear: number, minUnixTime: number, maxUnixTime: number): FiscalYearUnixTime {
        return new FiscalYearUnixTime(fiscalYear, minUnixTime, maxUnixTime);
    }
}

export const LANGUAGE_DEFAULT_FISCAL_YEAR_FORMAT_VALUE: number = 0;

export class FiscalYearFormat {
    private static readonly allInstances: FiscalYearFormat[] = [];
    private static readonly allInstancesByType: Record<number, FiscalYearFormat> = {};
    private static readonly allInstancesByTypeName: Record<string, FiscalYearFormat> = {};

    public static readonly StartYYYY_EndYYYY = new FiscalYearFormat(1, 'StartYYYY_EndYYYY');
    public static readonly StartYYYY_EndYY = new FiscalYearFormat(2, 'StartYYYY_EndYY');
    public static readonly StartYY_EndYY = new FiscalYearFormat(3, 'StartYY_EndYY');
    public static readonly EndYYYY = new FiscalYearFormat(4, 'EndYYYY');
    public static readonly EndYY = new FiscalYearFormat(5, 'EndYY');

    public static readonly Default = FiscalYearFormat.EndYYYY;

    public readonly type: number;
    public readonly typeName: string;

    private constructor(type: number, typeName: string) {
        this.type = type;
        this.typeName = typeName;

        FiscalYearFormat.allInstances.push(this);
        FiscalYearFormat.allInstancesByType[type] = this;
        FiscalYearFormat.allInstancesByTypeName[typeName] = this;
    }

    public static values(): FiscalYearFormat[] {
        return FiscalYearFormat.allInstances;
    }

    public static valueOf(type: number): FiscalYearFormat | undefined {
        return FiscalYearFormat.allInstancesByType[type];
    }

    public static parse(typeName: string): FiscalYearFormat | undefined {
        return FiscalYearFormat.allInstancesByTypeName[typeName];
    }
}

File: pkg/core/fiscalyear.go
package core

import (
	"fmt"

	"github.com/mayswind/ezbookkeeping/pkg/errs"
)

// FiscalYearStart represents the fiscal year start date as a uint16 (month: high byte, day: low byte)
type FiscalYearStart uint16

// Fiscal Year Start Date Type
const (
	FISCAL_YEAR_START_DEFAULT FiscalYearStart = 0x0101 // January 1
	FISCAL_YEAR_START_MIN     FiscalYearStart = 0x0101 // January 1
	FISCAL_YEAR_START_MAX     FiscalYearStart = 0x0C1F // December 31
	FISCAL_YEAR_START_INVALID FiscalYearStart = 0xFFFF // Invalid
)

var MONTH_MAX_DAYS = []uint8{
	uint8(31), // January
	uint8(28), // February (Disallow fiscal year start on leap day)
	uint8(31), // March
	uint8(30), // April
	uint8(31), // May
	uint8(30), // June
	uint8(31), // July
	uint8(31), // August
	uint8(30), // September
	uint8(31), // October
	uint8(30), // November
	uint8(31), // December
}

// NewFiscalYearStart creates a new FiscalYearStart from month and day values
func NewFiscalYearStart(month uint8, day uint8) (FiscalYearStart, error) {
	if !isValidFiscalYearMonthDay(month, day) {
		return 0, errs.ErrFormatInvalid
	}

	return FiscalYearStart(uint16(month)<<8 | uint16(day)), nil
}

// GetMonthDay extracts the month and day from FiscalYearType
func (f FiscalYearStart) GetMonthDay() (uint8, uint8, error) {
	if f < FISCAL_YEAR_START_MIN || f > FISCAL_YEAR_START_MAX {
		return 0, 0, errs.ErrFormatInvalid
	}

	// Extract month and day (month in high byte, day in low byte)
	month := uint8(f >> 8)
	day := uint8(f & 0xFF)

	if !isValidFiscalYearMonthDay(month, day) {
		return 0, 0, errs.ErrFormatInvalid
	}

	return month, day, nil
}

// String returns a string representation of FiscalYearStart in MM/DD format
func (f FiscalYearStart) String() string {
	month, day, err := f.GetMonthDay()

	if err != nil {
		return "Invalid"
	}

	return fmt.Sprintf("%02d-%02d", month, day)
}

// isValidFiscalYearMonthDay returns whether the specified month and day is valid
func isValidFiscalYearMonthDay(month uint8, day uint8) bool {
	return uint8(1) <= month && month <= uint8(12) && uint8(1) <= day && day <= MONTH_MAX_DAYS[int(month)-1]
}

// FiscalYearFormat represents the fiscal year format as a uint8
type FiscalYearFormat uint8

// Fiscal Year Format Type Name
const (
	FISCAL_YEAR_FORMAT_DEFAULT           FiscalYearFormat = 0
	FISCAL_YEAR_FORMAT_STARTYYYY_ENDYYYY FiscalYearFormat = 1
	FISCAL_YEAR_FORMAT_STARTYYYY_ENDYY   FiscalYearFormat = 2
	FISCAL_YEAR_FORMAT_STARTYY_ENDYY     FiscalYearFormat = 3
	FISCAL_YEAR_FORMAT_ENDYYYY           FiscalYearFormat = 4
	FISCAL_YEAR_FORMAT_ENDYY             FiscalYearFormat = 5
	FISCAL_YEAR_FORMAT_INVALID           FiscalYearFormat = 255 // Invalid
)

// String returns a textual representation of the long date format enum
func (f FiscalYearFormat) String() string {
	switch f {
	case FISCAL_YEAR_FORMAT_DEFAULT:
		return "Default"
	case FISCAL_YEAR_FORMAT_STARTYYYY_ENDYYYY:
		return "StartYYYY-EndYYYY"
	case FISCAL_YEAR_FORMAT_STARTYYYY_ENDYY:
		return "StartYYYY-EndYY"
	case FISCAL_YEAR_FORMAT_STARTYY_ENDYY:
		return "StartYY-EndYY"
	case FISCAL_YEAR_FORMAT_ENDYYYY:
		return "EndYYYY"
	case FISCAL_YEAR_FORMAT_ENDYY:
		return "EndYY"
	case FISCAL_YEAR_FORMAT_INVALID:
		return "Invalid"
	default:
		return fmt.Sprintf("Invalid(%d)", int(f))
	}
}
